### PR TITLE
Jet-Nemotron — EAGLE3 + Varlen Dynamic Conv

### PR DESCRIPTION
**Pull request by alex-t-hu**
## Motivation

Built off this PR: https://github.com/sgl-project/sglang/pull/12448
This PR makes the existing JetNemotron implementation faster, and implements speculative decoding for JetNemotron.

## Modifications

Topk>1 and topk=1 tree verification kernel for DynamicConv
Varlen prefill kernel for DynamicConv
Decode kernel for DynamicConv

## Accuracy Tests

1 RTX A6000, cuda graph on:
| Model + Inference Type                                                | GSM8K | MMLU |
|-----------------------------------------------------------------------|:-----:|:----:|
| Jet-Nemotron-2B                                                      | 0.763 | 0.622 |
| Jet-Nemotron-2B + EAGLE3 (topk=1, num_draft_tokens=6, num_steps=5)   | 0.766 | 0.621 |
| Jet-Nemotron-2B + EAGLE3 (topk=4, num_draft_tokens=8, num_steps=5)   | 0.757 | 0.621 |
| Jet-Nemotron-2B + EAGLE3 v2 (topk=1, num_draft_tokens=6, num_steps=5) | 0.787 | 0.606 |

| Model + Inference Type                                                | MMMU |
|-----------------------------------------------------------------------|:-----:|
| Jet-Nemotron-2B-vlm                                                      | 0.371 |
| Jet-Nemotron-2B-vlm + EAGLE3 (topk=4, num_draft_tokens=8, num_steps=5)   | 0.371 |

#### New implementation (no spec dec):
image

#### Original implementation:
image

### Speculative Decoding result
We trained an EAGLE3 checkpoint for jet-nemotron-2b-instruct and evaluated on mtbench
image

### CODE REVIEW

File: python/sglang/srt/layers/attention/jet_nemotron/dconv_step.py [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+            + pid_b * Cache_stride_b
+            + offs_d * Cache_stride_d
+            + 1 * Cache_stride_w
+        )
+        cache_val_1 = tl.load(cache_ptr_1, mask=d_mask, other=0.0)
+
+        # Accumulate Part 1
+        accumulator += cache_val_1 * k_val_0
+
+        # Cache Update: Store cache_val_1 -> cache_ptr_0 not needed
```
[the remark this answers is not in the document]
Comment (alex-t-hu) [reply python/sglang/srt/layers/attention/jet_nemotron/dconv_step.py]: that's ok

File: python/sglang/srt/models/jet_nemotron.py [hunk 2]
```diff
@@ -419,96 +279,124 @@ class JetNemotronDecoderLayer(nn.Module):
...
         self.quant_config = quant_config
-
-        self.model = Qwen2Model(
-            config,
-            quant_config=quant_config,
-            prefix=add_prefix("model", prefix),
-            decoder_layer_type=JetNemotronDecoderLayer,
-        )
-
+        self.model = JetNemotronModel(config, quant_config, prefix)
```
Comment (futrime) [inline python/sglang/srt/models/jet_nemotron.py]: Qwen2Model can directly be used.

**Comment by alex-t-hu** [comment]
thank you @attack204 

there's this bug when running speculative decoding v2?

https://github.com/sgl-project/sglang/issues/13352

so maybe we can just merge this in first? and then fix the bug for everyone later?

File: python/sglang/srt/speculative/eagle_worker.py [hunk 3]
```diff
@@ -752,9 +752,17 @@ def verify(self, batch: ScheduleBatch, spec_info: EagleVerifyInput):
...
             else:
                 max_relative_indices_per_req = accepted_length - 1
-            self.target_worker.model_runner.attn_backend.update_mamba_state_after_mtp_verify(
-                max_relative_indices_per_req, self.target_worker.model_runner.model
-            )
+            if (
+                self.target_worker.model_runner.jet_nemotron_config is not None
+                and spec_info.topk == 1
+            ):
+                self.target_worker.model_runner.attn_backend.update_jet_nemotron_topk1_state_after_mtp_verify(
```
Comment (yizhang2077) [inline python/sglang/srt/speculative/eagle_worker.py]: why jet_nemotron topk=1 logic is different? the same as `eagle_worker_v2.py`

File: python/sglang/srt/mem_cache/memory_pool.py [hunk 4]
```diff
@@ -205,20 +207,35 @@ def __init__(
...
-                    )
-                    for conv_shape in conv_state_shape
-                ]
+                if self.is_jet_nemotron_cache and speculative_eagle_topk == 1:
+                    intermediate_conv_window_cache = [
+                        torch.zeros(
+                            size=(
+                                num_mamba_layers,
+                                size + 1,
+                                speculative_num_draft_tokens + conv_shape[1] - 2,
```
Comment (yizhang2077) [inline python/sglang/srt/mem_cache/memory_pool.py:216]: why `topk=1` in `jet_nemotron` memory layout is special

File: python/sglang/srt/layers/attention/attention_registry.py [hunk 5]
```diff
@@ -206,7 +207,10 @@ def attn_backend_wrapper(runner: "ModelRunner", full_attn_backend: "AttentionBac
                     runner.server_args.attention_backend == "ascend"
                 ), "ascend backend is the only supported backend on NPU for hybrid GDN models, use --attention-backend ascend to specify the backend."
             logger.info(f"Using hybrid linear attention backend for hybrid GDN models.")
-            linear_attn_backend = GDNAttnBackend(runner)
+            if runner.jet_nemotron_config is not None:
+                linear_attn_backend = JetNemotronAttnBackend(runner)
```
Comment (yizhang2077) [inline python/sglang/srt/layers/attention/attention_registry.py:211]: maybe we could separate `jet_nemotron` out of `hybrid_gdn_config`?

File: python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py [hunk 6]
```diff
@@ -83,7 +85,7 @@ def _forward_metadata(self, forward_batch: ForwardBatch):
             query_start_loc = torch.arange(
                 0, bs + 1, dtype=torch.int32, device=self.device
             )
-        elif forward_batch.forward_mode.is_extend():
+        elif forward_batch.forward_mode.is_extend(include_draft_extend_v2=True):
```
Comment (yizhang2077) [inline python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py:88]: I am not sure if hybrid linear attention is compatible with eagle v2, do you have any idea? @hebiao064

File: python/sglang/srt/speculative/eagle_worker_v2.py [hunk 7]
```diff
@@ -706,6 +706,41 @@ def verify(
         else:
             verified_id = torch.empty((0,), device=self.device, dtype=torch.int32)
 
+        if self.target_worker.model_runner.hybrid_gdn_config is not None:
```
Comment (yizhang2077) [inline python/sglang/srt/speculative/eagle_worker_v2.py:710]: maybe we need to wrap it into another function?

**Review by yizhang2077 — COMMENTED** [review]
Could we add some ut for eagle3+`Jet-Nemotron`?

File: python/sglang/srt/speculative/eagle_worker.py (hunk 3 above)
Comment (alex-t-hu) [reply python/sglang/srt/speculative/eagle_worker.py]: i think for topk=1, we can make memory layout special to improve performance by reduce amount of memory transfer

File: python/sglang/srt/layers/attention/attention_registry.py (hunk 5 above)
Comment (alex-t-hu) [reply python/sglang/srt/layers/attention/attention_registry.py:211]: this could lead to more code changes since `hybrid_gdn_config` appears in a lot of places?

File: python/sglang/srt/speculative/eagle_worker_v2.py (hunk 7 above)
Comment (alex-t-hu) [reply python/sglang/srt/speculative/eagle_worker_v2.py:710]: nice yup did that
